protocol/superchainerc20/L1-native-factory.md

```diff
@@ -0,0 +1,166 @@
+
+```solidity
+mapping(address superc20 => address remoteToken) public deployments;
+```
+
+that maps each deployed `OptimismSuperchainERC20` address to the corresponding L1 address (remote address).
+
+## Implementation
+
+**Proxy Factory**
+
+Will follow the [`Proxy.sol` implementation](https://github.com/ethereum-optimism/optimism/blob/v1.1.4/packages/contracts-bedrock/src/universal/Proxy.sol) and `delegatecall()` to the factory implementation address.
+
+**Factory Implementation**
+
+```solidity
+contract OptimismSuperchainERC20Factory is ISemver {
+  mapping(address superc20 => address remoteToken) public deployments;
+  string public constant version = "1.0.0";
+
+  event OptimismSuperchainERC20Deployed(address indexed superchainERC20, address indexed remoteToken, string name, string symbol, uint8 decimals);
+
+  constructor() {}
+
+  function deploy(address _remoteToken, string memory _name, string memory _symbol, uint8 _decimals) external returns (address _superchainERC20) {
+    bytes memory _creationCode = abi.encodePacked(
+        type(BeaconProxy).creationCode,
+        abi.encode(Predeploy.OptimismSuperchainERC20Beacon, abi.encodeCall(SUPERCHAIN_IMPL.initialize.selector, (_remoteToken, _name, _symbol, _decimals)))
+    );
+
+    bytes32 salt = keccak256(abi.encode(_remoteToken, _name, _symbol, _decimals));
+    _superchainERC20 = CREATE3.deploy(salt, _creationCode);
+
+    deployments[_superchainERC20] = _remoteToken;
+
+    emit OptimismSuperchainERC20Deployed(_superchainERC20, _remoteToken _name, _symbol, _decimals);
+  }
+}
+```
+
+The name `_remoteToken` was chosen over `_l1Token` to contemplate future L2 native cases.
+
+**BeaconProxy (OptimismSuperchainERC20)**
+
+The Token should follow a simple BeaconProxy implementation, like Openzeppelin:
+
+https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/beacon/BeaconProxy.sol
+
+**Beacon Contract**
+
+```solidity
+contract OptimismSuperchainERC20Beacon is IBeacon {
+  /// TODO: Replace with real implementation address
+  address internal constant IMPLEMENTATION_ADDRESS = 0x0000000000000000000000000000000000000000;
+
+  function implementation() external pure override returns (address) {
+    return IMPLEMENTATION_ADDRESS;
+  }
+}
+```
+
+**Implementation**
+The implementation should include
+
+- An `initialize` function that takes `(address _remoteToken, string _name, string _symbol, uint8 decimals)` and stores these in the storage of the BeaconProxy.
+- A function to reinitialize after proxy deployments or upgrades.
+
+## User Stories
+
+### `OptimismSuperchainERC20` deployment
+
+1. Anyone can deploy a `OptimismSuperchainERC20` by calling the `deploy` function in the `OptimismSuperchainERC20Factory` contract with `_remoteToken`, `_name`, `_symbol` and `decimals` as inputs.
+2. The factory stores the `_remoteToken` address for each deployed SuperchainERC20 address.
+
+```mermaid
+sequenceDiagram
```

Comment:
Since we're deploying an OptimismSuperchainERC20 BeaconProxy, should this diagram also have an `OptimismSuperchainERC20BeaconProxy.initialize()` call? Or is there nothing that needs to be initialized?

Reply:
correct